### PR TITLE
Add feature to automatically remove supervisor based on retention period

### PR DESCRIPTION
**Pull request by maytasm**
Add feature to automatically remove supervisors based on retention period

### Description

We currently already have tasklog auto cleanup (https://github.com/apache/druid/pull/3677) and audit logs auto cleanup (https://github.com/apache/druid/pull/11084). This PR adds a similar auto cleanup based on duration (time to retained) but for the supervisor table to auto clean up terminated supervisors

This is useful when Druid user has a high churn of task / datasource in a short amount of time causing the metadata store size to grow uncontrollably. 

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: docs/configuration/index.md [hunk 1]
```diff
@@ -747,6 +747,9 @@ These Coordinator static configurations can be defined in the `coordinator/runti
...
 |--------|-----------|---------|-------|
 |`druid.coordinator.period.metadataStoreManagementPeriod`|How often to run metadata management tasks in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. |No | `PT1H`|
+|`druid.coordinator.kill.supervisor.on`| Boolean value for whether to enable automatic deletion of terminated supervisors. If set to true, Coordinator will periodically remove terminated supervisors from the supervisor table in metadata storage.| No | False| 
+|`druid.coordinator.kill.supervisor.period`| How often to do automatic deletion of terminated supervisor in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. Value must be greater than `druid.coordinator.period.metadataStoreManagementPeriod`. Only applies if `druid.coordinator.kill.supervisor.on` is set to True.| No| `P1D`|
+|`druid.coordinator.kill.supervisor.durationToRetain`| Duration of terminated supervisor to be retained from created time in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. Only applies if `druid.coordinator.kill.supervisor.on` is set to True.| Yes if `druid.coordinator.kill.supervisor.on` is set to True| None|
```
Comment (loquisgon) [inline docs/configuration/index.md:752]: The code has 1S for default: PT-1s

Comment (maytasm) [reply docs/configuration/index.md:752]: The default in the code is used to fail cluster startup if the value is not set by the user. Basically code check for the value being positive (meaning that the user must set it and override the default value). For documentation simplicity sake, the default value in the code is the same effect as None and that the user must set the value.

File: docs/configuration/index.md [hunk 2]
```diff
@@ -747,6 +747,9 @@ These Coordinator static configurations can be defined in the `coordinator/runti
 |Property|Description|Required?|Default|
 |--------|-----------|---------|-------|
 |`druid.coordinator.period.metadataStoreManagementPeriod`|How often to run metadata management tasks in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. |No | `PT1H`|
+|`druid.coordinator.kill.supervisor.on`| Boolean value for whether to enable automatic deletion of terminated supervisors. If set to true, Coordinator will periodically remove terminated supervisors from the supervisor table in metadata storage.| No | False| 
+|`druid.coordinator.kill.supervisor.period`| How often to do automatic deletion of terminated supervisor in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. Value must be greater than `druid.coordinator.period.metadataStoreManagementPeriod`. Only applies if `druid.coordinator.kill.supervisor.on` is set to True.| No| `P1D`|
```
Comment (abhishekagarwal87) [inline docs/configuration/index.md:751]:
```suggestion
|`druid.coordinator.kill.supervisor.period`| How often to do automatic deletion of terminated supervisor in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. Value must be equal to or greater than `druid.coordinator.period.metadataStoreManagementPeriod`. Only applies if `druid.coordinator.kill.supervisor.on` is set to "True".| No| `P1D`|
```

File: docs/configuration/index.md (hunk 1 above)
Comment (abhishekagarwal87) [inline docs/configuration/index.md:752]:
```suggestion
|`druid.coordinator.kill.supervisor.durationToRetain`| Duration of terminated supervisor to be retained from created time in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) duration format. Only applies if `druid.coordinator.kill.supervisor.on` is set to "True".| Yes if `druid.coordinator.kill.supervisor.on` is set to "True"| None|
```

File: docs/operations/metrics.md [hunk 3]
```diff
@@ -256,6 +256,7 @@ These metrics are for the Druid Coordinator and are reset each time the Coordina
 |`interval/skipCompact/count`|Total number of intervals of this datasource that are skipped (not eligible for auto compaction) by the auto compaction.|datasource.|Varies.|
 |`coordinator/time`|Approximate Coordinator duty runtime in milliseconds. The duty dimension is the string alias of the Duty that is being run.|duty.|Varies.|
 |`coordinator/global/time`|Approximate runtime of a full coordination cycle in milliseconds. The `dutyGroup` dimension indicates what type of coordination this run was. i.e. Historical Management vs Indexing|`dutyGroup`|Varies.|
+|`metadata/kill/supervisor/count`|Total number of terminated supervisors automatically deleted from metadata store supervisor table per each Coordinator kill supervisor duty run. This metric can help adjust `druid.coordinator.kill.supervisor.durationToRetain` configuration based on if more or less terminated supervisors need to be deleted per cycle. Note that this metric is only emitted when `druid.coordinator.kill.supervisor.on` is set to true.| |Varies.|
```
Comment (abhishekagarwal87) [inline docs/operations/metrics.md:259]:
```suggestion
|`metadata/kill/supervisor/count`|Total number of terminated supervisors that were automatically deleted from metadata store per each Coordinator kill supervisor duty run. This metric can help adjust `druid.coordinator.kill.supervisor.durationToRetain` configuration based on whether more or less terminated supervisors need to be deleted per cycle. Note that this metric is only emitted when `druid.coordinator.kill.supervisor.on` is set to true.| |Varies.|
```

File: server/src/main/java/org/apache/druid/metadata/MetadataSupervisorManager.java [hunk 4]
```diff
@@ -34,4 +34,12 @@
   Map<String, List<VersionedSupervisorSpec>> getAll();
 
   Map<String, SupervisorSpec> getLatest();
+
+  /**
+   * Remove terminated supervisor created older than the given timestamp.
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/metadata/MetadataSupervisorManager.java:39]:
```suggestion
   * Remove terminated supervisors created before the given timestamp.
```

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java [hunk 5]
```diff
@@ -249,6 +252,34 @@ public Void withHandle(Handle handle) throws Exception
...
 
+  @Override
+  public int removeTerminatedSupervisorsOlderThan(long timestamp)
+  {
+    int removedCount = 0;
+    DateTime dateTime = DateTimes.utc(timestamp);
+    Map<String, SupervisorSpec> supervisors = getLatest();
+    for (Map.Entry<String, SupervisorSpec> supervisor : supervisors.entrySet()) {
+      final SupervisorSpec spec = supervisor.getValue();
+      if (spec instanceof NoopSupervisorSpec) {
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java:263]: does this mean that it is a terminated supervisor? can you add a comment clarifying this bit

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java [hunk 6]
```diff
@@ -249,6 +252,34 @@ public Void withHandle(Handle handle) throws Exception
...
+        removedCount += dbi.withHandle(
+            handle -> {
+              Update sql = handle.createStatement(
+                  StringUtils.format(
+                      "DELETE FROM %1$s WHERE spec_id = :spec_id AND created_date < :date_time",
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java:268]: does it make sense to delete in batch than doing one by one which will be slow?

File: server/src/main/java/org/apache/druid/server/coordinator/duty/KillSupervisors.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+import org.apache.druid.metadata.MetadataSupervisorManager;
+import org.apache.druid.server.coordinator.DruidCoordinatorConfig;
+import org.apache.druid.server.coordinator.DruidCoordinatorRuntimeParams;
+
+public class KillSupervisors implements CoordinatorDuty
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/KillSupervisors.java:31]: javadocs

File: docs/configuration/index.md (hunk 1 above)
Comment (loquisgon) [reply docs/configuration/index.md:752]: ok, LGTM

File: docs/configuration/index.md (hunk 2 above)
Comment (maytasm) [reply docs/configuration/index.md:751]: Done

File: docs/configuration/index.md (hunk 1 above)
Comment (maytasm) [reply docs/configuration/index.md:752]: Done

File: docs/operations/metrics.md (hunk 3 above)
Comment (maytasm) [reply docs/operations/metrics.md:259]: Done

File: server/src/main/java/org/apache/druid/metadata/MetadataSupervisorManager.java (hunk 4 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/metadata/MetadataSupervisorManager.java:39]: Done

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java (hunk 5 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java:263]: Yep. Terminated supervisor will have it's latest supervisorSpec as NoopSupervisorSpec. (Basically NoopSupervisorSpec is used as a tombstone indicator)

File: server/src/main/java/org/apache/druid/server/coordinator/duty/KillSupervisors.java (hunk 7 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/KillSupervisors.java:31]: Done

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java (hunk 6 above)
Comment (maytasm) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataSupervisorManager.java:268]: Done

**Review by abhishekagarwal87 — APPROVED** [review]
LGTM @maytasm. Does this PR need a "release notes" label?

**Pull request merged**